### PR TITLE
8255231: Avoid upcalls when initializing the statSampler

### PR DESCRIPTION
**Pull request by cl4es**
Current implementation of the statSampler does upcalls to System.getProperty to collect values for a number of properties that are all provided by the VM itself. And since the sampling starts before any user code run then no property can have changed.

I suggest refactoring the code so that no upcalls are made normally - while asserting this invariant holds using assert-only upcalls. 

This is a small startup optimization - reducing the startup sequence by approx. 300k instructions and 70k branches in my linux-x64 setup.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x32 | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8255231](https://bugs.openjdk.java.net/browse/JDK-8255231): Avoid upcalls when initializing the statSampler

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 6e2202277964f1d0a23268f6bf64897adbb7ac2c
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/802/head:pull/802`
`$ git checkout pull/802`

### CODE REVIEW

File: src/hotspot/share/runtime/statSampler.cpp [hunk 1]
```diff
@@ -215,76 +214,58 @@ const char* StatSampler::get_system_property(const char* name, TRAPS) {
  * creates other property based instrumentation conditionally.
  */
 
-// stable interface, supported counters
-static const char* property_counters_ss[] = {
+// stable interface, supported counters in the JAVA_PROPERTY name space
```
Comment (iklam) [inline src/hotspot/share/runtime/statSampler.cpp:217]:
```
 * The list of System Properties that have corresponding PerfData
 * string instrumentation created by retrieving the named property's
 * value from System.getProperty() and unconditionally creating a
 * PerfStringConstant object initialized to the retrieved value. This
 * is not an exhaustive list of Java properties with corresponding string
 * instrumentation as the create_system_property_instrumentation() method
 * creates other property based instrumentation conditionally.
```
I found the above comment unreadable. Is it possible to clarify it?

**Review by iklam — APPROVED** [review]
Looks good to me. A nice clean up.

File: src/hotspot/share/runtime/statSampler.cpp (hunk 1 above)
Comment (cl4es) [reply src/hotspot/share/runtime/statSampler.cpp]: I reworked it a bit to get rid of the static array and keep comments more coherent. I think it reads a bit better now.

**Review by iklam — APPROVED** [review]
LGTM

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
So just to be clear, the statSampler was doing a System.getProperty() upcall, for a property that was previously set by the VM, where the value was initially obtained via a read of the fields in java.lang.VersionProps and stored in the VM in either JDK_Version of VM_version. So now you just use that JDK_Version/VM_Version value directly. And to allow that you had to add in the reading of the "short version" java.version property.

This seems okay in principle. I have a couple of nits:
- can we refer to JDK_Version::java_version instead of "short_version"? I see where the short comes from in VersionProps, but as it represents the java.version value I think we can just use that in the VM for clarity.
- are we concerned that when we store these values in JDK_Version/VM_Version we are artificially constraining their length?
- now that java.version is being read from VersionProps can you please add a comment to that effect in the template:

+ // This field is read by HotSpot
    private static final String java_version =
        "@@VERSION_SHORT@@";

Thanks,
David

**Review by dholmes-ora — APPROVED** [review]
Thanks for making the suggested changes.

I think we need a further RFE to add some error checking for the sizes of the various property strings in relation to the fixed size arrays that have been allocated to them.

Thanks,
David

**Comment by cl4es** [comment]
@dholmes @iklam: thanks for reviewing!

/integrate

**Pull request closed without merging**